### PR TITLE
Update repo attributes on each build

### PR DESCRIPTION
**Pull request by arunagw**
The idea is to keep repo attributes updated always so we can track public/private flag change on repo from GitHub.

### CODE REVIEW

File: app/jobs/buildable.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 module Buildable
   def perform(payload_data)
     payload = Payload.new(payload_data)
+
+    RepoUpdater.new(payload).run
```
Comment (arunagw) [inline app/jobs/buildable.rb]: I am not sure what to call this. This class will mostly be handling repo updates and deactivating repo based on conditions.

File: app/jobs/buildable.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
 module Buildable
   def perform(payload_data)
     payload = Payload.new(payload_data)
+
+    repo_updater = RepoUpdater.new(payload)
```
Comment (gylaz) [inline app/jobs/buildable.rb]: How about we get away from the "Land of Nouns" and call this `UpdateRepoStatus`?

Comment (arunagw) [inline app/jobs/buildable.rb]: Yes. Will update this.

**Comment by gylaz** [comment]
@arunagw Do you mind adding a descriptive commit message about why something was done?

File: app/services/update_repo_status.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  def repo_attributes
+    {
+      full_github_name: payload.full_repo_name,
+      private: payload.private_repo?,
+    }
+  end
+
+  def deactive_repo(repo)
+    if repo.previous_changes[:private] == change_public_to_private
+      repo.deactivate
```
Comment (gylaz) [inline app/services/update_repo_status.rb]: I just went through some of the repos that went from public to private, and manually reached out to folks. One person converted and started paying. Based on that, I'm not sure it's a good idea to disable the repo right away. I think it might be best to track such repos (maybe via Administrate dashboard), and reach out to them separately.

Thoughts @thoughtbot/hound?

File: app/jobs/buildable.rb [hunk 4]
```diff
@@ -1,6 +1,8 @@
 module Buildable
   def perform(payload_data)
     payload = Payload.new(payload_data)
+
+    UpdateRepoStatus.new(payload).run
```
Comment (gylaz) [inline app/jobs/buildable.rb]: What do you think about `BuildRunner` calling this? This way we wouldn't need to find `repo` in two places.

File: spec/models/payload_spec.rb [hunk 5]
```diff
@@ -150,6 +150,34 @@
...
+    context "when repo is private" do
+      it "returns true" do
+        payload_json = {
+          "repository" => {
+            "private" => true,
+          }
+        }
+        payload = Payload.new(payload_json)
+
+        expect(payload.private_repo?).to be_truthy
```
Comment (gylaz) [inline spec/models/payload_spec.rb]: What do you think about `expect(payload.private_repo?).to eq true`?

File: spec/models/payload_spec.rb [hunk 6]
```diff
@@ -150,6 +150,34 @@
...
+    context "when repo is public" do
+      it "returns false" do
+        payload_json = {
+          "repository" => {
+            "private" => false,
+          }
+        }
+        payload = Payload.new(payload_json)
+
+        expect(payload.private_repo?).to be_falsy
```
Comment (gylaz) [inline spec/models/payload_spec.rb]: What do you think about `expect(payload.private_repo?).to eq false`?

**Comment by gylaz** [comment]
@arunagw I've rebased latest master and force pushed.

File: app/services/update_repo_status.rb (hunk 3 above)
Comment (thorncp) [inline app/services/update_repo_status.rb]: Could also make it easy to deactivate those repos in Administrate, if we don't get a response. :+1:

File: app/jobs/buildable.rb (hunk 4 above)
Comment (arunagw) [inline app/jobs/buildable.rb]: I think finding repo is fine at two places because it's anyway in background. Previously we were updating repo when finding that in `BuildRunner`.

I think updating repo needs to come before `BuildRunner` so we make sure that it's updated.

What you think?

File: app/services/update_repo_status.rb (hunk 3 above)
Comment (arunagw) [inline app/services/update_repo_status.rb]: We can also add another flag like `pending_deactivation` and that can generate a report to know.

We can avoid deactivation in this PR and just update a new status and then we can do another PR for report or admin handle.

File: app/jobs/buildable.rb (hunk 4 above)
Comment (gylaz) [inline app/jobs/buildable.rb]: Sounds good.

File: app/services/update_repo_status.rb (hunk 3 above)
Comment (gylaz) [inline app/services/update_repo_status.rb]: There's a fairly easy query we can run to see which repos are private, but don't have a subscription (and are not bulk customers). So, an additional flag might be unnecessary.

Comment (arunagw) [inline app/services/update_repo_status.rb]: Should we not deactivate then? And just change status based on GitHub payload?

Comment (gylaz) [inline app/services/update_repo_status.rb]: I'm leaning towards not activating, then we can run a query (or add an admin dashboard) to see who's not paying.

Comment (arunagw) [inline app/services/update_repo_status.rb]: Taken care here https://github.com/thoughtbot/hound/commit/1d200eebd3f57a95916e8aaef45ec337dc93a7dd 

I think you meant not deactivating right?

**Comment by arunagw** [comment]
In this PR we are just updating repo attributes whenever a new build comes in.

File: app/services/update_repo_status.rb (hunk 3 above)
Comment (gylaz) [inline app/services/update_repo_status.rb]: Yep.

File: app/services/update_repo_status.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+class UpdateRepoStatus
+  pattr_initialize :payload
+
+  def run
+    if repo
+      repo.update_attributes(repo_attributes)
```
Comment (gylaz) [inline app/services/update_repo_status.rb]: What do you think about using just `update` (`repo.update(repo_attributes)`)?

**Comment by gylaz** [comment]
:+1:

File: spec/jobs/buildable_spec.rb [hunk 8]
```diff
@@ -23,6 +18,19 @@ class BuildableTestJob < ActiveJob::Base
       expect(BuildRunner).to have_received(:new).with(payload)
       expect(build_runner).to have_received(:run)
     end
+
+    it "runs repo updater" do
+      repo_updater = double(:repo_updater, run: nil)
```
Comment (teoljungberg) [inline spec/jobs/buildable_spec.rb]: We've consistently used strings as the names for the `double`, what do you think about doing so here too?

File: spec/services/update_repo_status_spec.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      expected_repo_name = "foo/bar"
+      repo = create(:repo, :active)
+      payload = double(
+        "Payload",
+        github_repo_id: repo.github_id,
+        full_repo_name: expected_repo_name,
+        private_repo?: repo.private,
+      )
+
+      UpdateRepoStatus.new(payload).run
```
Comment (teoljungberg) [inline spec/services/update_repo_status_spec.rb]: What do you think about having the `repo` being returned by this service? That would avoid the `.reload` call and make the code a bit clearer, I think.

**Comment by teoljungberg** [comment]
A few ideas and comments, but this looks good to me

File: spec/services/update_repo_status_spec.rb (hunk 9 above)
Comment (arunagw) [inline spec/services/update_repo_status_spec.rb]: I usually don't prefer anything to be returned from a `.run` method. But what if repo not found? It might return nil.

File: spec/jobs/buildable_spec.rb (hunk 8 above)
Comment (arunagw) [inline spec/jobs/buildable_spec.rb]: Yes.

**Comment by arunagw** [comment]
Will merge this in when CI is green. Thanks for reviews.
